### PR TITLE
feat: mean pooling

### PR DESCRIPTION
**Pull request by triandco**
The current implementation does not include [pooling](https://github.com/qdrant/fastembed/blob/9c72d2f59f91f87753da07c12a6f1082e233ecb3/fastembed/text/pooled_embedding.py#L50). This result in wildly different embedding to fastembed python counterpart. 

This pull request is an attempt to ensure the default embedding output get an accuracy up to 1-e6.

This pull request includes:
- [x] Added a hard coded embedding test, 
- [x] Added mean_pool function, 
- [x] Applied mean_pool in embedding generation.

### CODE REVIEW

File: src/tests.rs [hunk 1]
```diff
@@ -283,3 +283,123 @@ fn test_user_defined_reranking_model() {
     assert_eq!(results.len(), documents.len());
     assert_eq!(results.first().unwrap().index, 0);
 }
+
+
+#[test]
+fn compare_python_embeddings() {
+    let model = TextEmbedding::try_new(InitOptions {
+        model_name: EmbeddingModel::AllMiniLML6V2,
```
Comment (Anush008) [inline src/tests.rs:291]: Hey @triandco.
Can you please check for `bge-small-en-1.5`?

Comment (Anush008) [reply src/tests.rs:291]: I'd expect this to be a not applicable to all models.
`bge-small-en-v1.5`'s embeddings were in parity.

Comment (triandco) [reply src/tests.rs:291]: Yes you are right, bge-small-en-1.5 only use pooling_mode_cls_token.

i’ll extend this with a pooling config loading so we can match each model settings.

Comment (Anush008) [reply src/tests.rs:291]: Also, for a per model post process step, if needed, you can refer to https://github.com/Anush008/fastembed-rs/blob/main/src%2Fmodels%2Fsparse.rs

Used like [so](https://github.com/Anush008/fastembed-rs/blob/main/src%2Fsparse_text_embedding.rs#L236).

Comment (Anush008) [reply src/tests.rs:291]: Also, please add yourself to the authors list in the `Cargo.toml` file.

Comment (triandco) [reply src/tests.rs:291]: Thank you @Anush008, I’ve added the todo to the PR description. I’ll work through it in the next couple of days.

Comment (Anush008) [reply src/tests.rs:291]: 🙏

Comment (Anush008) [reply src/tests.rs:291]: Hi @triandco.

TLDR from #102.

@triandco please don't bother looking into the changes in #102. You can continue working on this PR as is.

Will merge this first and then adapt #102 accordingly.

File: src/models/text_embedding.rs [hunk 2]
```diff
@@ -257,3 +261,56 @@ pub fn models_list() -> Vec<ModelInfo<EmbeddingModel>> {
...
+    pub fn load_pooling_config(path: &PathBuf) -> Result<PoolingConfig, LoadPoolingError>{
+        let file = File::open(path).map_err(|_| LoadPoolingError::FailToLoadCacheConfigFile)?;
+        let reader = BufReader::new(file);
+        serde_json::from_reader::<_,PoolingConfig>(reader).map_err(|_| LoadPoolingError::FailToDeserialiseConfig)
+    }
+
+    // Pick a default pooling method for a model because our embed method only output a single set of vector
+    pub fn best_pooling_method(config: PoolingConfig) -> Pooling {
+        if config.pooling_mode_cls_token { Pooling::Cls }
+        else if config.pooling_mode_mean_tokens { Pooling::Mean }
```
Comment (Anush008) [inline src/models/text_embedding.rs:313]: This could benefit from a match mapping.

File: src/tests.rs [hunk 3]
```diff
@@ -283,3 +286,820 @@ fn test_user_defined_reranking_model() {
...
+        -1.08737415e-02,
+        3.26007046e-02,
+        -3.86322886e-02,
+        -1.04548130e-02,
+        4.71166559e-02,
+        1.69442664e-03,
+        -9.29024518e-02,
+        8.17656144e-02,
+        5.08985445e-02,
+    ];
```
Comment (Anush008) [inline src/tests.rs:692]: How about we check only the first 5 elements?
Seems unnecessary to type these many literals here.

File: src/tests.rs [hunk 4]
```diff
@@ -283,3 +286,820 @@ fn test_user_defined_reranking_model() {
...
+        6.72190730e-03,
+        2.01416835e-02,
+        -4.58171926e-02,
+        2.07999609e-02,
+        3.75747420e-02,
+        8.51716474e-02,
+        -7.08120465e-02,
+        2.96163429e-02,
+        4.38641012e-02,
+    ];
```
Comment (Anush008) [inline src/tests.rs:1098]: See https://github.com/Anush008/fastembed-rs/pull/99/files#r1713350588

File: src/text_embedding.rs [hunk 5]
```diff
@@ -138,14 +141,23 @@ impl TextEmbedding {
                 .expect("Failed to retrieve model.onnx_data.");
         }
 
+        // prioritise loading pooling config if available, if not (thanks qdrant!), look for it in hardcoded
+        let post_processing = match model_repo.get("1_Pooling/config.json") {
+            Err(_) => model_name.get_default_pooling_method(),
```
Comment (Anush008) [inline src/text_embedding.rs:146]: Hey @triandco.
If we're specifying the pooling strategy for each model in `get_default_pooling_method()`, is trying to read the config file needed at all?

File: src/models/text_embedding.rs (hunk 2 above)
Comment (triandco) [reply src/models/text_embedding.rs]: I thought about it, but it is kinda messy with a bunch of boolean. If we are to add max, sum and meansqrt, it might look mighty ugly. So I kept the if else for now. ^^;

File: src/text_embedding.rs (hunk 5 above)
Comment (triandco) [reply src/text_embedding.rs:146]: the default method is only a fall back. Ideally, we should always get it from the source in online mode, since it can get updated. 

I only added the default one because qdrant did a good job of uploading the onnx without a pooling config. ^^;

Down side is more code, more behaviour, more maintainance. Happy to remove it if you feel it actually make the code harder to maintain.

Comment (Anush008) [reply src/text_embedding.rs:146]: Right now the switch in `get_default_pooling_method()` doesn't have a catch all clause. So every model must be specified.
So, I guess, reading the config file won't be necessary and we can specify it when adding the model.

Comment (triandco) [reply src/text_embedding.rs]: no worries, it's gone now.

File: src/pooling.rs [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+pub enum LoadPoolingError {
+    FailToLoadCacheConfigFile,
+    FailToDeserialiseConfig,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum Pooling {
+    Cls,
+    Mean,
+}
```
[the remark this answers is not in the document]
Comment (triandco) [reply src/pooling.rs:7]: My intention with none None currently is that there is no default pooling, so we fall back to the Pooling:Cls. 

If we were to add Pooling::None, it means that we shouldn't pool the output at all, In such case, I will update its behaviour in text_embedding.rs line 337 to force the ort output into a 2 dimension vec for return or throw if the shape is mismatch. What do you think @Anush008 ?

[the remark this answers is not in the document]
Comment (Anush008) [reply src/pooling.rs:7]: > If we were to add Pooling::None, it means that we shouldn't pool the output at all, In such case, I will update its behaviour in text_embedding.rs line 337 to force the ort output into a 2 dimension vec for return or throw if the shape is mismatch. What do you think @Anush008 ?

TBH,  I have no preference, since I don't have expertise in ML. I am happy to go with your preference.

[the remark this answers is not in the document]
Comment (triandco) [reply src/pooling.rs:7]: Let's leave it for now. I would only be confident forcing the tensor shape down to vector with good tests. I haven't come across a model that has built in pooling yet. So it hard to write a sensible tests. 

I can easily add it later if we find one. I'll leave the todo in there for now.

**Review by Anush008 — APPROVED** [review]
Thank you @triandco. You'll see the feature in v4.

cc @denwong47 . This is the final state of the PR.

**Comment by denwong47** [comment]
Noted, will take a look after work.
